_posts/function-contracts.md

```diff
@@ -0,0 +1,405 @@
+
+```rust
+fn gcd(mut max: u64, mut min: u64) -> u64 {
+    if min > max {
+        std::mem::swap(&mut max, &mut min);
+    }
+
+    let rest = max % min;
+    if rest == 0 { min } else { gcd(min, rest) }
+}
+
+```
+
+This is by far the costliest part of the code of `TokenBucket::new`. The rest of the function was a fixed set of divisions, but here we have a division and a recursive call back to `gcd`. It is not immediately obvious how busy this computation is but in the [worst case](https://en.wikipedia.org/wiki/Euclidean_algorithm#Worst-case) the number of steps (recursions) approaches 1.5 times the number of bits needed to represent the input numbers. Meaning that for two large 64 bit numbers it can take almost 96 iterations for a single call to `gcd`. Recall that our smallest input to `TokenBucket::new` is a non-deterministic value in the range $0< x < 18446744073709$ (up to 45 non-zero bits). If a harness that uses `gcd` does not heavily constrain the input, Kani would have to unroll the recursion at least 68 times and then execute it symbolically. An expensive operation.
+
+Since `gcd` is the most expensive part of verifying this harness it is an ideal target for using a contract to make it more efficient. In fact by replacing it with a contract we not only make this harness more efficient but in fact any harness that uses `gcd` directly or indirectly (for instance by calling `TokenBucket::new`. In firecracker for instance there are three more harnesses that all call `gcd` and thus benefit from modular verification.
+
+## Introducing a Function Contract
+
+Function contracts are conceptually rather simple actually. They comprise a set of conditions which characterize the behavior of the function, similar to the conditions you would use in a test case. There are different types of conditions which are also called the *clauses* of the contract. The first type of clause we will introduce here is the `ensures` clause.
+
+```rust
+#[kani::ensures(max % result == 0 && min % result == 0 && result != 0)]
+fn gcd(mut max: u64, mut min: u64) -> u64 {
+    if min > max {
+        std::mem::swap(&mut max, &mut min);
+    }
+
+    let rest = max % min;
+    if rest == 0 { min } else { gcd(min, rest) }
+}
+
+```
+
+The `ensures` clause describes the relationship of the return of the function with the arguments that the function was called with. It is often also called a *postcondition*, because it is a *condition* that must hold after (*post*) the execution of the function. With Kani the contents of the `ensures` clause can be any Rust expression that returns `bool`. However the expression may not allocate, deallocate or modify heap memory or perform I/O. A single function may have multiple `ensures` clauses which functions as though they had ben joined with `&&`. Our example could thus also have been written as
```

Comment:
```suggestion
The `ensures` clause describes the relationship of the return of the function (i.e., `result`) with the arguments that the function was called with. It is often also called a *postcondition*, because it is a *condition* that must hold after (*post*) the execution of the function. With Kani the contents of the `ensures` clause can be any Rust expression that returns `bool`. However the expression may not allocate, deallocate or modify heap memory or perform I/O. A single function may have multiple `ensures` clauses which functions as though they had been joined with `&&`. Our example could thus also have been written as
```